Developed Test Cases for Lambdas

JDaniel41 opened the pull request:
# Summary
I created some test cases in `test_register_user.py` and `test_log_visit.py` to check the functionality of the lambda. These test cases are automated using `pytest`. However, to successfully automate these test cases, I had to make some changes to the lambdas. These involved refactoring the code itself in the lambdas. I also had to move around the lambdas so I could import these lambdas in our test cases.

## Test Cases
In `cdk/visit/lambda_code/test_log_visit.py`, I imported `pytest` and `moto`. These two libraries have been added to the `requirements.txt` file. `Pytest` is used to automate the test cases in `test_log_visit.py` which are copied from the AWS console. These test cases hit mocked versions of our dynamoDB table and our Simple Email Service. This way, we do not mess with our production tools.

In order to run these test cases, I also had to move the lambdas into their own subdirectories. This way, we can import them without the test file being uploaded to the AWS console. This change inadvertently resolves #45 since now the files aren't uploaded repeatedly.

## Lambda Refactoring
Next, I refactored the lambda code in `log_visit.py` to accept mocked versions of the table. To do this, I added optional parameters that default to the production versions of the AWS services. This way, in production, the code will continue to hit the production services. Meanwhile, when we run our test cases, we can pass mocked versions of the database and simple email service. Similar changes will be needed to `register_user.py`.

## Testing the Changes
First, I ran `cdk deploy` to ensure that the lambdas are still uploaded to the console. They are uploaded correctly and API Gateway is still configured to hit them. Next, I ran the test cases locally using `pytest` in the `lambda_code` folder. Both test cases pass.

JDaniel41 committed:
Statrted Test File

JDaniel41 committed:
Added pytest and moto to requirements.txt

JDaniel41 commented:
Currently running into an issue where when I run `pytest` in the directory, I get this error. Is this something with my requirements.txt file @ophilli @dggorsen ?

```
_ ERROR collecting visit/lambda_code/test_log_visit.py _
ImportError while importing test module '/Users/jonathanmdaniel/Documents/Code/Clemson/unified-makerspace/cdk/visit/lambda_code/test_log_visit.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/opt/homebrew/Cellar/python@3.9/3.9.10/Frameworks/Python.framework/Versions/3.9/lib/python3.9/importlib/__init__.py:127: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
test_log_visit.py:2: in <module>
    from aws_cdk import (
E   ImportError: cannot import name 'Stack' from 'aws_cdk' (unknown location)
```

ophilli commented on cdk/visit/lambda_code/test_log_visit.py:
```diff
@@ -0,0 +1,96 @@
+from log_visit import *
+from aws_cdk import (
+    Stack,
+    aws_lambda as _lambda,
+    assertions
+)
+from moto import mock_dynamodb2
```
Why are you importing these here? It doesn't look like you ever use them below.

JDaniel41 replied:
`mock_dynamodb2` is meant to be used for locally mocking a Dynamo DB table. It gets used as a decorator on line 20. My assumption is that it is neccesary so the calls don't mess with a dynamo db table in prod.

`aws_cdk` imports were from trying to follow a tutorial I found. Working on trying to figure out if that is still neccesary.

JDaniel41 replied:
The `aws_cdk` imports are unnecessary.

JDaniel41 committed:
Refactor log_visit to accept parameterized tables

I modified the `checkRegistration()`, `registrationWorkflow()`, `addVisitEntry()` and `handler()` functions in `log_visit.py` to take extra parameters. These parameters are meant to make the log_visit lambda unit-testable. These changes do not change the production behavior, as running the code in the AWS console will produce the same results.

JDaniel41 committed:
First Round of Unit Tests

JDaniel41 committed:
Moved Register User and Log Visit.py into subfolders of lambda_code

JDaniel41 committed:
Updated Test Case

JDaniel41 commented:
Updated the code to be more organized. I also added testing directions and noted that these changes also now solve #45

JDaniel41 committed:
Added Test Cases for Register User and Refactored Lambda

JDaniel41 committed:
Moved Environment Variable Declaration to Python Script

JDaniel41 commented:
- Added Test Cases for `register_user.py` and updated the lambda to follow the same convention as `log_visit.py`

JDaniel41 committed:
Added GitHub Action for Pytest

JDaniel41 committed:
Updated Python Version to 3.9

JDaniel41 commented:
These test cases can be ran by the Github Action in #82

JDaniel41 committed:
Update Requirements.txt to include pytest and moto

dggorsen commented on cdk/visit/lambda_code/log_visit/log_visit.py:
```diff
@@ -1,5 +1,6 @@
 import json
 import datetime
+from xmlrpc import client
```
This feels like its likely the wrong import. I would expect the client type we are referring to to live in the boto package.

JDaniel41 replied:
Removed the import since it was unused.

dggorsen commented on cdk/visit/lambda_code/register_user/register_user.py:
```diff
@@ -34,7 +34,7 @@ def addUserInfo(user_info):
 
 
-def handler(request, context):
+def handler(request, context, table=users):
```
Have you considered an object oriented approach to this? e.g.

```
class RegisterUserFunction():
    def __init__(self, users):
        if table is None:
            # Get the service resource.
            dynamodb = boto3.resource('dynamodb')
            # Get the table name.
            TABLE_NAME = os.environ["TABLE_NAME"]
            # Get table objects
            self.users = dynamodb.Table(TABLE_NAME)
        else: 
            self.users = users

    def handle_register_user_request(self, request, context):
        # implement business logic here

fn = RegisterUserFunction(users)

# Entry point of our lambda function remains the same, but we now have a handler object that 
# encapsulates the logic necessary to register the user
handler(request, context):
  fn.handle_register_user_request(request, context)
```

Through these means you should be able to leave the rest of the function interfaces as-is and your tests can be written against the RegisterUserFunction class methods.

JDaniel41 replied:
I did not originally. At the time, I wanted to go with an approach that minimized the changes to the handler function and the approach I took was the first one that came to mind. I definitely do see the merit in the approach you describe though. @ophilli What are your thoughts? I'm 50/50 on which approach I prefer.

JDaniel41 replied:
Just implemented it in the latest commit!

JDaniel41 committed:
Updated requirements.txt

JDaniel41 committed:
Added Region Environment Variable to Github Action

JDaniel41 committed:
Added Default Region Environment Variable to Github Action

JDaniel41 committed:
Updated Action and Removed Unused Import

JDaniel41 commented:
- Folded the GitHub Actions into this PR.

JDaniel41 commented:
- Also just tested the most recent changes in the AWS Lambda console and they still function.

JDaniel41 committed:
Refactored Lambdas to use Object-Oriented Approach

JDaniel41 committed:
Clean-up Changes

JDaniel41 commented:
- Changed lambdas to use the object-oriented style. I tested that the production versions work by using `cdk deploy`.

ophilli commented:
This branch cannot be rebased due to conflicts. Can you please fix the conflicts?

https://stackoverflow.com/questions/50012036/github-cant-rebase-my-feature-branch-this-branch-cannot-be-rebased-due-to-con

dggorsen commented on cdk/visit/lambda_code/log_visit/log_visit.py:
```diff
@@ -0,0 +1,196 @@
+
+        if client is None:
+            AWS_REGION = os.environ['AWS_REGION']
+            self.client = boto3.client('ses', region_name=AWS_REGION)
```
can we make this member variable more descriptive? e.g. `ses_client`

JDaniel41 replied:
Done!

dggorsen commented on cdk/visit/lambda_code/log_visit/log_visit.py:
```diff
@@ -0,0 +1,196 @@
+def handler(request, context):
+    # This will be hit in prod, and will connect to the stood-up dynamodb
+    # and Simple Email Service clients.
+    return LogVisitFunction(None, None).handle_log_visit_request(request, context)
```
Small performance tip:
We can create this function object during lambda's static initialization phase, thereby only instantiating our clients once when the function is initialized. We can then reuse the clients that we have created throughout the lifetime of our lambda's execution environment

See https://docs.aws.amazon.com/lambda/latest/operatorguide/static-initialization.html for more detailse.g.

Ex.
```
# Initialize once.
log_visit_function = LogVisitFunction(None, None)

# Lambda entry point which is executed on every lambda invocation
def handler(request, context):
    return log_visit_function.handle_log_visit_request(request, context)
```

For a little context, the way that lambda works when it is invoked:
1. Creates an "execution environment" which effectively creates an ephemeral container (lasting ~15 mins) capable of being invoked `n` times over those 15 minutes. 
2. Invokes the function by running your "handler" code
3. The next time the function is invoked it will attempt to reuse a "warm" execution environement

The initialization period of an "execution environment"  is an expensive process relative to invoking a function that already has a "warm" execution environment. You can look into this process by googling "lambda cold start."

JDaniel41 replied:
Sounds good! I went ahead and moved around the code to run during the initialization process.

dggorsen reviewed (changes requested):
Great job! Couple very small comments for performance purposes. This is great though :D

JDaniel41 committed:
Initialized Functions during initialization period and renamed SES client

JDaniel41 committed:
Updated Test Case Code to import files AFTER we set environemnt variables

JDaniel41 commented:
- Rebased the Code to avoid conflicts to address @ophilli notes.
- Moved the initialization code so the lambdas are more efficient on startup.

dggorsen reviewed (approved):
Nice work!
